### PR TITLE
 Remove guard_size_oblivious from vector_norm decomposition.

### PR DESCRIPTION
**Pull request by laithsakka**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148809

This PR remove the usage of guard_size_oblivious in vector_norm by inlining it in the runtime check, 
this prevent any data dependent error from ever appearing here at the locations where guard_size_oblivious
used to exist. Before this PR it used to break potentially. This is NOT BC breaking or changing of semantics from eager. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Comment by bobrenjc93** [comment]
Can you add a test? Especially for these first few migrations it’d be good to ensure it’s doing what we expect.

File: torch/_refs/linalg/__init__.py [hunk 1]
```diff
@@ -107,29 +108,36 @@ def vector_norm(
...
-                f"dimension {d} because this dimension is empty and the "
-                "operation does not have an identity",
+                sym_or(x.numel() != 0, dim is not None and len(dim) != 0),
+                lambda: f"linalg.vector_norm cannot compute the {ord} norm on an empty tensor "
+                "because the operation does not have an identity",
             )
+            shape = x.shape
+            assert (
+                dim is not None
+            )  # mypy does not seem to be able to see through check?
```
Comment (aorenste) [inline torch/_refs/linalg/__init__.py:133]: It's true that mypy doesn't understand torch._check() - but also your check isn't guaranteeing that dim is not None, is it? (If `x.numel() != 0` then `dim` could be `None`)

Comment (aorenste) [inline torch/_refs/linalg/__init__.py:133]: I'm sure pyfmt reflowed this onto three lines...
```suggestion
            # mypy does not seem to be able to see through check?
            assert dim is not None
```

File: torch/_refs/linalg/__init__.py [hunk 2]
```diff
@@ -61,6 +61,7 @@ def _check_norm_dtype(dtype: Optional[torch.dtype], x_dtype: torch.dtype, fn_nam
 
 
 import operator
+from operator import or_ as sym_or
```
[the remark this answers is not in the document]
Comment (laithsakka) [reply torch/_refs/linalg/__init__.py:64]: i just want to call operator.or_ but want it named sym_or cause its what would happen.

File: torch/_refs/linalg/__init__.py [hunk 3]
```diff
@@ -107,29 +108,36 @@ def vector_norm(
...
+        # x.numel() != 0.
+        # x.numel() != 0 is repeated in the checks on purpose to handle the situation
+        # where x.numel() == 0 is data dependent. In such case it will be checked at runtime.
+        if guard_or_true(x.numel() == 0):
             torch._check(
-                shape[d] != 0,
-                lambda: f"linalg.vector_norm cannot compute the {ord} norm on the "
-                f"dimension {d} because this dimension is empty and the "
-                "operation does not have an identity",
+                sym_or(x.numel() != 0, dim is not None and len(dim) != 0),
```
[the remark this answers is not in the document]
Comment (laithsakka) [reply torch/_refs/linalg/__init__.py]: will force guarding.

File: torch/_refs/linalg/__init__.py (hunk 1 above)
Comment (laithsakka) [reply torch/_refs/linalg/__init__.py]: this is exiting code i did not add this

File: test/test_linalg.py [hunk 4]
```diff
@@ -1445,6 +1447,46 @@ def run_test_case(input, ord, dim, keepdim, norm_dtype):
...
+                return tensor_unbacked_size
+
+        # When expect_runtime_check we expect a runtime check to inserted.
+        def test(ord: Union[float, int], dim: Optional[DimsType], expect_runtime_check:bool)->None:
+            m = Mod(ord, dim)
+            exported_program: torch.export.ExportedProgram = torch.export.export(
+                m , args= tuple(torch.tensor([1])))
+            print(exported_program.graph_module.code.strip())     
+
+        # # Those are situation where we should throw error if input.size()==0
```
Comment (eellison) [inline test/test_linalg.py:1475]: Why do we have these commented out ? if its for something that should work, but doesnt, you can add a test with expected failure.

**Review by eellison — COMMENTED** [review]
Some of the decomps are for numerics, which I think we should guard for. see http://github.com/pytorch/pytorch/pull/144073/

cc @isuruf

File: torch/_refs/linalg/__init__.py (hunk 1 above)
Comment (laithsakka) [reply torch/_refs/linalg/__init__.py]: but anyway i am changing this will add an if

Comment (laithsakka) [reply torch/_refs/linalg/__init__.py]: i am removing this assertion will add if instead.

File: test/test_linalg.py (hunk 4 above)
Comment (laithsakka) [reply test/test_linalg.py:1475]: ah this is draft i am adding unit tests for got to mark the PR draft

File: torch/_refs/linalg/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (laithsakka) [reply torch/_refs/linalg/__init__.py]: you can try this example
```
class Mod(torch.nn.Module):
    def forward(self,a: torch.Tensor, b: torch.Tensor):
        x = a.item()
        torch._check(x>10 or True)

        # torch._check(operator.or_( x>10 , True))
        return b*10
    

a = torch.tensor([1])
b = torch.torch.ones(10, 10, device="cuda",dtype=torch.float64) 

exported_program: torch.export.ExportedProgram = torch.export.export(
    Mod(), args=(a,b))
print(exported_program.graph_module.code.strip())
```

File: torch/_refs/linalg/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (laithsakka) [reply torch/_refs/linalg/__init__.py:64]: i can add sym_or = operator._or to the symbolic shapes and add a comment about it for future clairty

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -1277,6 +1277,12 @@ def statically_known_true(x: Union[bool, SymBool]) -> bool:
     return x
 
 
+# a or b is eager, meaning that (a==1 or True) will specialize a.
+# but operator.or_(a==1, True) generates sympy expression and avoid
+# the specialization.
+sym_or = operator.or_
```
Comment (bobrenjc93) [inline torch/fx/experimental/symbolic_shapes.py:1283]: this already exists on SymNode https://www.internalfb.com/code/fbsource/[6f1893cbc5666e46b21f179a39bea4d5df2fd462]/xplat/caffe2/torch/fx/experimental/sym_node.py?lines=439

File: torch/fx/experimental/symbolic_shapes.py [hunk 6]
```diff
@@ -1277,6 +1277,12 @@ def statically_known_true(x: Union[bool, SymBool]) -> bool:
     return x
 
 
+# a or b is eager, meaning that (a==1 or True) will specialize a.
+# but operator.or_(a==1, True) generates sympy expression and avoid
+# the specialization.
```
Comment (bobrenjc93) [inline torch/fx/experimental/symbolic_shapes.py:1282]: It took me some time to understand this. The key is that operator.or_ will respect the dunder method `__or__` on symnode instead of eagerly evaluating. Perhaps you can update the comment to reflect this?

File: test/test_linalg.py [hunk 7]
```diff
@@ -1445,6 +1447,61 @@ def run_test_case(input, ord, dim, keepdim, norm_dtype):
...
+                expect_index_0_check,
+            )
+
+        # dim is int.
+        test(-1, 1, True)
+
+        # dim is None
+        test(-1, None, True)
+
+        # len(dim) ==0
```
Comment (bobrenjc93) [inline test/test_linalg.py:1493]: s/==0/== 0

File: test/test_linalg.py [hunk 8]
```diff
@@ -1445,6 +1447,61 @@ def run_test_case(input, ord, dim, keepdim, norm_dtype):
...
+        test(-1, [], True)
+
+        # shape[d] == 0
+        test(-1, [0], False, True)
+
+        # u0+1==0 is False so we do not see a runtime assert
```
Comment (bobrenjc93) [inline test/test_linalg.py:1499]: s/u0+1==0/u0 + 1 == 0

Comment (bobrenjc93) [reply test/test_linalg.py:1499]: s/so we do not see a runtime assert/so we do not expect to fail on the runtime assert

**Comment by bobrenjc93** [comment]
Overall looks good. One thing that'd be helpful for reviewers on these migration PRs is to include summaries with a sense of the spirit and risk of the change. Are we introducing any BC changes? What new checks did we add and why? Do we expect more data-dependent errors or less? etc. It might also make sense to add comments in the meta functions themselves so future readers have an easier time understanding the intention behind these guards/checks.

cc @pianpwk

File: torch/_refs/linalg/__init__.py [hunk 9]
```diff
@@ -97,6 +97,34 @@ def diagonal(
...
+
+    if not (ord < 0.0 or ord == float("inf")):
+        return
+
+    torch._check(
+        sym_or(
+            x.numel() != 0,
+            not isinstance(dim, IntLike) and dim is not None and len(dim) != 0,
+        ),
+        "linalg.vector_norm cannot compute the {ord} norm on an empty tensor "
```
Comment (aorenste) [inline torch/_refs/linalg/__init__.py:113]:
```suggestion
        lambda: (f"linalg.vector_norm cannot compute the {ord} norm on an empty tensor "
```

File: torch/_refs/linalg/__init__.py [hunk 10]
```diff
@@ -97,6 +97,34 @@ def diagonal(
...
+    )
+
+    shape = x.shape
+    if dim is not None and not isinstance(dim, IntLike):
+        for d in dim:
+            torch._check(
+                sym_or(x.numel() != 0, d < len(shape) and d >= 0 and shape[d] != 0),
+                "linalg.vector_norm cannot compute the {ord} norm on the "
+                f"dimension {d} because this dimension is empty and the "
+                "operation does not have an identity",
```
Comment (aorenste) [inline torch/_refs/linalg/__init__.py:124]: You don't want to use an f-string outside of a lambda (because it will compute the string even if the check passes)

```suggestion
                lambda: (f"linalg.vector_norm cannot compute the {ord} norm on the "
                f"dimension {d} because this dimension is empty and the "
                "operation does not have an identity"),
```

File: torch/fx/experimental/symbolic_shapes.py (hunk 6 above)
Comment (aorenste) [reply torch/fx/experimental/symbolic_shapes.py:1282]: Agreed - Definitely need to explain the difference between `or` and `operator.or_` for symnodes...

File: torch/_refs/linalg/__init__.py (hunk 10 above)
Comment (laithsakka) [reply torch/_refs/linalg/__init__.py:124]: this is intentional torch.compile does not support lambda strings here. 
https://github.com/pytorch/pytorch/issues/149997

File: torch/_refs/linalg/__init__.py (hunk 9 above)
Comment (laithsakka) [reply torch/_refs/linalg/__init__.py:113]: this is intentional torch.compile does not support lambda strings here.
https://github.com/pytorch/pytorch/issues/149997

File: torch/_refs/linalg/__init__.py (hunk 10 above)
Comment (laithsakka) [reply torch/_refs/linalg/__init__.py:124]: what we can do , is have a utility that pre check to see if the compile check will be deferred or not, and then calls this or that, or fix the issue above, buts its not related to this change. 
previoulsy it was ok, because we would never defer those runtime asserts. not since we can defer them we hit this limitation.

File: test/test_linalg.py (hunk 8 above)
Comment (laithsakka) [reply test/test_linalg.py:1499]: mm i meant we do not see runtime assert in the generated graph which was i this testing

File: torch/fx/experimental/symbolic_shapes.py (hunk 6 above)
Comment (laithsakka) [reply torch/fx/experimental/symbolic_shapes.py]: i will add.

File: test/test_linalg.py [hunk 11]
```diff
@@ -1445,6 +1447,61 @@ def run_test_case(input, ord, dim, keepdim, norm_dtype):
...
+                in exported_program.graph_module.code,
+                expect_numel_runtime_check,
+            )
+            self.assertEqual(
+                "Runtime assertion failed for expression Ne(u0, 0) | Ne(u0*(u0 + 1)*(u0 + 2), 0)"
+                in exported_program.graph_module.code,
+                expect_index_0_check,
+            )
+
+        # dim is int.
```
Comment (bobrenjc93) [inline test/test_linalg.py:1487]: remove the `.` or add punctuation to the rest?

File: test/test_linalg.py [hunk 12]
```diff
@@ -1445,6 +1447,61 @@ def run_test_case(input, ord, dim, keepdim, norm_dtype):
...
+        # dim is None
+        test(-1, None, True)
+
+        # len(dim) == 0
+        test(-1, [], True)
+
+        # shape[d] == 0
+        test(-1, [0], False, True)
+
+        # u0+1 == 0 is False we do not see a runtime assert in the generated graph.
```
Comment (bobrenjc93) [inline test/test_linalg.py:1499]: `u0 + 1`, notice the spaces

File: torch/fx/experimental/symbolic_shapes.py [hunk 13]
```diff
@@ -1277,6 +1277,19 @@ def statically_known_true(x: Union[bool, SymBool]) -> bool:
...
+# a data dependent error for an expression “if u0==1 or True”. or over guarding for
+# “if s0==1 or True”.
+
+# On the other hand, when we use operator.or_, then dynamo will generate
+# a sympy expression Sympy.Or(u0==1, True) without evaluating the args first.
+
+# When the whole expression is passed to evaluation in that case, we do not throw a
+# data dependent error or guard because we can statically know the result is True
+# before unpacking the symbols.
+sym_or = operator.or_
```
Comment (bobrenjc93) [inline torch/fx/experimental/symbolic_shapes.py:1290]: given that @pian is introducing sym_or in https://github.com/pytorch/pytorch/pull/150456#event-17082477541, does it make sense to rebase on top of his PR?

**Review by bobrenjc93 — APPROVED** [review]
